### ISSUE TITLE
MCP-0009: Undefined Modifications

### ISSUE THREAD

**Issue by modelica-trac-importer** [tbeu/ModelicaSpecification#468]
**Reported by kurzbach on 7 Dec 2013 18:07 UTC**
There is a new MCPI which tries to solve the first open issue ("Handling of default values not solved") from the MCPI-0008_CustomAnnotations(#1293) by introducing a "undefined" keyword in modifications. As a side effect it adds to Modelica the possibility to remove existing modifications from parameters and variables and, to replace them by a more sophisticated computation.

The document is stored in SVN:
https://svn.modelica.org/projects/MCP-Ideas/MAinternal/MCPI-0009_UndefinedModifications

In its current state, it is basically a summary of the mailings at the modelica design list, mostly from Martin Otter, Michael Tiller and me(Gerd Kurzbach).

Please use this ticket to comment on this MCPI.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1377

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 9 Dec 2013 14:36 UTC**
I have several concerns that I've expressed via email, but I wanted to make sure they are logged here.

The first concern is that if I define something and then "undefine" it, I should end up back where I started.  From the email discussions, it seems that this is not currently the case with this proposal.

Second, I think the semantics for annotations and non-annotations should be the same.  It will be confusing if we represent annotations as records and then have the effect of undefined be different when applied to a record depending on whether that record is being used as an annotation or not.

Third, as I understand the proposal so far, if I apply undefined to an annotation the entry in the annotation "disappears".  This I find rather alarming.  To me, I found the benefit of using records in annotations was the ability to constrain the structure of the annotation so I knew exactly what data would be present there.  If we can "erase" entries, it seems we are right back where we were before we started talking about custom annotation (where annotations are just free form "dictionaries" with no assurances).  It seems to me that if we want different fields we should undefine the value of one field but we should simply define a different record with different fields.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 10 Dec 2013 05:45 UTC**
The concerns might be addressed by allowing unused parameters to be undefined. For example:
```
model UndefinedParameter
  parameter Real p1;
  parameter Real p2 = 5;
  Real x(start = 0);
equation 
  der(x) = p2;
end UndefinedParameter;
```
Instead of calling the model invalid, the parameter p1 could be removed during model instantiation. Then the semantics would be the same for annotations and non-annotations. Undefining the value of a parameter later on would have the same effect as never defining it.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 24 Jan 2014 12:14 UTC**
There was a discussion between the authors of the Custom Annotations paper about the chapter "Optional Values", where the semantics of "undefined" was described. (23/24th January 2014)
Finally it was decided within a phone conference to remove the whole chapter from the paper, because it was not completely clear yet.

The relevant part is the following:

**Original paper Text:**
(Seven rules to apply "undefined")
 (1) If an identifier of a variable or parameter with undefined value (meaning whose binding equation has undefined as right hand side) occurs in an expression E1 (operator, function call, array index, etc.), the expression E1 is being replaced by undefined. If expression E1 is part of an expression E2, the rule is applied directly to E2.
  Example
   parameter Integer a = undefined;
   b = 2\* a; -> b = undefined;
   array[a] = 12; -> undefined = 12;
 (2) undefined must only occur on the right hand side of equations (and then form the right hand side of an equation). If it occurs at any other place (left-hand side, if-statement, etc.) an error is raised during the translation.
  Example
   Correct: parameter b = undefined;
   Wrong: if undefined then […]
 (3) An equation where undefined forms the right-hand-side must be an equation of a parameter declaration, or a binding equation in a modifier referring to a parameter or an equation within an annotation statement.
  Correct Examples:
   parameter b = undefined;
   extends A(x = undefined);
   annotation( OptimOptions(method = undefined))
  Wrong Examples:
   model […] equation b = undefined; […]
 (4) Rules (1), (2), and (3) are applied recursively until all indirect references to undefined elements have been resolved or an error occurred.
 (5) During translation of a simulation model, all declarations whose values are undefined are re-moved from the flattened equations (they are not used in the model, otherwise rule(2) or (3) would have caused an error)
 (6) Within annotation statements: the equations that contain undefined are removed.
 (7) Clean-up within annotations: Empty constructors or modifiers are being removed. This rule is applied recursively.

**Comment by Gerd:**

•	in rule (1) the text is fine, but in the examples you have two lines with equations, not modifications.  These are somehow unrelated, because the originally thinking was to only apply this rule to modifications and not to general equations. 
“undefined” is not an expression and cannot and should not occur in equations.  Basically if it would, it would hinder the second use case, just to remove modifications from lower levels and replace them by initial equations, or more sophisticated calculations.
•	in rule (2) the word equation should be replaced by modification. Any other use results in an syntax error.
•	rule (3) is somehow the repetition of (2)
•	rule (4): Basically only rule (1) is applied recursively, rule (2) isn’t something to apply.
•	rule (5) should be: “… all declarations which are “undefined”, and if their parameters or variables are not used in the model, are removed”. If they are used in equations (this is allowed), other equations or algorithms has to be provided to  give them a value. If not, an error occurs, complaining about a missing equation.
•	rule (6): “undefined” can only be in modifications, so all modifications (not only in annotations) containing it are removed.
•	rule (7): empty constructors cannot occur, because they are expressions and are already removed in rule (1). Empty modifications should be cleaned up.
•	The sentence after the rules: “undefined” does not create undefined variables or parameters, it just removes their modifications. And then, if they are not used anymore in the model, they are removed completely.
 
**Comment by Dirk:**
I have designed the rules so that it actually only works for applying parameters in modifications or annotations.
 
Nevertheless, when I write such rules, I _have_ to define what happens when the user chooses to use undefined somewhere else or when he uses a parameter that has been marked as undefined someplace else. Hence this rules, they are there for completion and to make clear when an error shall be raised.
 
If undefined shall only work in modifications (and no place else) then also Martin has misunderstood your proposal. Since his examples used undefined parameters and that was the whole point of all the use-cases. Anyway when you allow a parameter to be specified as undefined then users will immediately play around with it. It wouldn’t make sense to be super-super-restrictive since many of our potential users would immediately start to complain about all the restrictions. Then it is better not to use undefined at all.

**Comment by Gerd:**
> Nevertheless, when I write such rules, I _have_ to define what happens when the user chooses to use undefined somewhere else or when he uses a parameter that has been marked as undefined someplace else. 

This is the basic misunderstanding: The parameter is not marked as undefined, but the modification is removed and it is allowed to remove the parameter if it is not further used.

> Anyway when you allow a parameter to be specified as undefined then users will immediately play around with it. It wouldn’t make sense to be super-super-restrictive since many of our potential users would immediately start to complain about all the restrictions. 

Which restrictions?  There are no further restrictions. Instead it is allowed to use parameters which have a removed “undefined” modification inside of equations. And they are not really undefined, because the user can define another way to provide a value for them.
To handle undefined in expressions in equations is a complete different thing and not in the scope of this proposal. This would lead to the handling of undefined values at run time, because not all cases are decidable at compile time. This was discussed at the last design meeting and in the mailing list and is not wanted now, because it complicates the things so much. Nevertheless the current proposal might be extended in that direction in the future.

> Then it is better not to use undefined at all.

This was a requirement from the Custom Annotation proposal, so I’m not the real inventer of it.

**Comment by Martin:**
to my understanding, the rules introduced by Dirk have exactly the same effect as the previous (intended, but probably incomplete) rules. If not, maybe you can point out where there is a deviation. As I understand, the description with propagation of "undefined" in expressions (rule 1) and allowing "undefined" only at particular places (rule 3) is just another way to state that an "undefined" parameter is not allowed to be used in "regular" expressions, but only in modifiers and in annotations.

**Comment by Gerd:**
the problem arises, when parameters or variables which have an undefined modification are used in model equations and are replaced by the undefined expression. This suggest the examples in the paper but this is not intended at all.
It would prevent the second use case, to calculate their value using another way then a modification.

In rule (1) only the examples are misleading, the text is fine. Rule (2) continuous the wrong direction by using the term equation instead of modification. Then rule (2) and (3) say the same, so they can combined to one rule.
In rule (5) the wrong direction is formulated, so that the second use case is not possible anymore.

**Comment by Martin:**
I just read now the complete (modified) section about optional values. Unfortunately, you removed the last part with the two examples. I think these are important because they show other uses cases of "undefined" that are very useful. Please, re-introduce these two examples.

The previous version of the rules have been:
1.	The “undefined” element is not allowed in functions.
2.	When an „undefined“ element is referenced directly or indirectly in an annotation (such as annotation(OptimOptions(method = optimOptions.method))), then this annotation is removed.
3.	When „undefined“ is used as modifier, then the binding equation to the „undefined“ element is removed at this place (see example below).
4.	After performing (2) and (3) on all elements of a simulation model or block (so the model or block to be translated), all elements declared as “undefined” are removed, provided a binding equation was not provided in a modifier.

I think "rule (1)" is still useful, because "undefined" cannot be removed in functions and therefore it is not possible to used it in functions. It seems better to state this directly, in order to deduce it implicitly by other rules.
rule(2) is of course not good. It states a processing of the translator and has nothing to do with the language.
rule(4) is again something for the translator not for the language, so this is also not formulated in a good way.

So, to summarize, I agree that the rules needs to be improved. Due to the comments from Gerd, the proposal from Dirk seems to be also not yet o.k.

Here is another try:
1.	The “undefined” element is not allowed in functions.
2.	When „undefined“ is used as modifier, then the binding equation to the „undefined“ element is removed at this place (see example below).
3.	During translation of a simulation model or block the following actions are performed:
1.	Annotations that are "undefined" (= the modifier is defined as "undefined" or the modifier is a variable and this variable depends directly or indirectly on a variable declared as "undefined) are removed from the flattened equations.
2.	All declarations whose values are "undefined" (after applying all modifications) are removed from the flattened equations.
3.	Clean-up within annotations: Empty constructors or modifiers are being removed. This rule is applied recursively
Note, if a variable is declared as "undefined" and this variable is used in an equation, then during flattening an error occurs (so the model is invalid), because the declaration of the "undefined" variable was removed, and then the equation references a variable that is not declared.
To my understanding, from the language point of view the only needed rules are rule (1) and (2). All other rules are related to the translation phase.

**Comment by Dirk:**
> I just read now the complete (modified) section about optional values. Unfortunately, you removed the last part with the two examples. I think these are important because they show other uses cases of "undefined" that are very useful. Please, re-introduce these two examples.

I’ll see how I can reintroduce the examples but I also think it is good to keep the paper in scope. If the discussion on undefined leads to no consensus, it is probably the best to eradicate it from the paper completely. It seems to be a pretty immature concept right now.

> The previous version of the rules have been:
> 1.	The “undefined” element is not allowed in functions.
> 2.	When an „undefined“ element is referenced directly or indirectly in an annotation (such as annotation(OptimOptions(method = optimOptions.method))), then this annotation is removed.
> 3.	When „undefined“ is used as modifier, then the binding equation to the „undefined“ element is removed at this place (see example below).
> 4.	After performing (2) and (3) on all elements of a simulation model or block (so the model or block to be translated), all elements declared as “undefined” are removed, provided a binding equation was not provided in a modifier.

> I think "rule (1)" is still useful, because "undefined" cannot be removed in functions and therefore it is not possible to used it in functions. It seems better to state this directly, in order to deduce it implicitly by other rules.

No problem with that. It is indeed meaningful to add.

> rule(2) is of course not good. It states a processing of the translator and has nothing to do with the language.
> rule(4) is again something for the translator not for the language, so this is also not formulated in a good way.
> So, to summarize, I agree that the rules needs to be improved. Due to the comments from Gerd, the proposal from Dirk seems to be also not yet o.k.
> Here is another try:
> 1.	The “undefined” element is not allowed in functions.
> 2.	When „undefined“ is used as modifier, then the binding equation to the „undefined“ element is removed at this place (see example below).
> 3.	During translation of a simulation model or block the following actions are performed:
> 1.	Annotations that are "undefined" (= the modifier is defined as "undefined" or the modifier is a variable and this variable depends directly or indirectly on a variable declared as "undefined) are removed from the flattened equations.

That was my problem before: What does that precisely mean: “depends directly or indirectly on a variable declared as "undefined”. 
This is what I wanted to clarify with my rules. 
When you allow indirect dependence, users will immediately come up with use-cases like this:
```
parameter Real diameter  = undefined;
final parameter radius = diameter/2;
annotation(Tuner(min = -radius, max = radius))
```
I see no reason to forbid this and my rules handle this case in a clear and non-ambiguous way. (here, diameter, radius and the annotation would vanish) 
On the contrary: with “undefined” we add a new keyword and a new concept to the language. If we do that we should make it worth the effort. Introducing all the complexity but then forbidding everything is no good for simplifying the language. If we do it, we should do it so that the new concept can also be used for a wide range of application otherwise it is not worth the effort and better should be kept away from the language.  We have already enough of this little hacks in the language.

**Comment by Martin:**
> I’ll see how I can reintroduce the examples but I also think it is good to keep the paper in scope. If the discussion on undefined leads to no consensus, it is probably the best to eradicate it from the paper completely. It seems to be a pretty immature concept right now.

We are talking since 10 years to introduce something like "undefined" in Modelica because there are several use cases where this is needed (I am pretty sure that there are also several unclosed tickets on this topic). Its the first time that there is a reasonable proposal available. Therefore, if any possible I would like to have it in the paper. Furthermore, for custom annotations it will become really clumsy in practical every-day-use, if such a concept would not be available.

**Comment by Gerd:**
this is my try:


 (1)	If a modification of a variable or parameter depends directly or indirectly from a variable or parameter with “undefined” modification (meaning whose binding equation has undefined as right hand side) then this modification is replaced by an “undefined” modification. Other equations than modifications kept unchanged.
  Example:
  ```
   parameter Integer a = undefined; 
   annotation (b = 2* a); -> annotation (b = undefined); 
   annotation (c=if 3*b>15 then 1 else 0); -> annotation (c= undefined)
   equation if  a> 5 then …  ->  equation if  a> 5 then // unchanged
   ```
 (2)	The then the declarations having an “undefined” modification and  are not used in any other equation (they are not referenced by any equation) are removed including the modification from the equations.
 (3)	Then all remaining “undefined” modifications are removed.
 (4)	Clean-up within annotations: Empty modifiers are being removed. This rule is applied recursively. 

I don’t understand completely, why it shouldn’t be allowed in functions. It could be used to remove default arguments, so they have to be given explicitly.

Dirk, your example 
```
parameter Real diameter  = undefined;
final parameter radius = diameter/2;
annotation(Tuner(min = -radius, max = radius))
```
is perfectly fine and a user is allowed to do so. Finally all of them are removed. 

A slightly extended example:
```
model A
parameter Real diameter  = undefined;
final parameter radius = diameter/2;
annotation(Tuner(min = -radius, max = radius))
end A;

model B
     extends A(diameter(fixed=false));
     parameter Real sqare;
     initial equation 
           square =f(diameter); // assuming this cannot be solved symbolically for diameter
end B;
```
This would be allowed to do and all parameters and annotations are kept, because diameter is referenced in an equation.

One question remains: How to provide the values for the annotation in this case?

**Comment by Dirk:**
Ok then we agree on what we want to express, but this is what I expressed with my rules too.

The difference between our formulations is just the approach: I start with general rules and then enforce constraints (this is why (2) and (3) are not redundant), you make constraints (partly implicit) first and then state some rules on the constrained set. Well both is possible but it was unclear before for me when to generate errors.

Maybe it is the best if we both sleep one night over it and then talk tomorrow.

**Comment by Martin:**
I did not think about this application. But you are right, this is useful and needed. It would be good to add this example also to the paper to make the proposal more clear.

Gerd, you wrote: "if a modification of a variable or parameter depends directly or indirectly from a variable or parameter with “undefined” modification ". Dirk is right, that such a formulation is not clear and can be misleading (as in our previous formulations). Could you go through Dirks proposal again under this view and see whether your missing feature can be added (or improve your proposal, so that the "depend directly or indirectly" is precisely defined).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 24 Jan 2014 12:47 UTC**
I don't think the rules cited from the paper were clear. I believe there is a proposal above by Martin that is closer to what we want.

My main problem with the rules are rule (1):
If expression E1 is part of an expression E2, the rule is applied directly to E2. 

Assume we have:
```
parameter Boolean b=false;
parameter Boolean b2=undefined;
parameter Boolean b3=b and b2;
```
In Dymola [and other tools] we normally perform constant folding for `false and …` -> `false` (which is currently allowed but not required by the spec). So is b3 false or undefined? Similar issues with if-brances etc.

Having undefined take precedence would mean that we need an additional pass in the symbolic processing to remove all undefined; and would also be confusing for users. 
The reason I write confusing for users is that we already have a similar issue caused by if-brances for parameter-bindings. It also requires a much stricter "pass"-handling in translators than is currently needed - which makes the cost non-trivial for implementors.

Having false take precedence would significantly alter the description – since it would imply that we first perform symbolic processing where “undefined” is kept and then sort of “pull”. That would correspond to treating a parameter equal to undefined as if it had no value (or at least as it had no literal value); tools already need to handle parameters without literal value.

In addition I don’t really see why we should removed undefined parameters from the model in general.

I can understand removing undefined custom-annotations and possibly have some special _kind_ of parameters that are only intended for custom-annotations – but when I think of:
```
parameter Real mass=density*r^3; 
parameter Real density=undefined;
...
equation
 mass*der(v)=f;
...
``` 
– I don't see that making mass (which is used in equations) disappear is a good goal.
I can see issues, e.g. how do we generate good diagnostics for these cases?
But I cannot see the main benefits.

I can also understand the benefit of using e.g. `extends A(mass=undefined)` to remove a binding equation for mass.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 24 Jan 2014 13:03 UTC**
Dear Hans,
could you have a look at my last comment in the discussion, please. There are 4 points. 
What do you think about that?
Thanks, 
Gerd.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 24 Jan 2014 13:15 UTC**
Replying to [comment:5 kurzbach]:
> Dear Hans,
> could you have a look at my last comment in the discussion, please. There are 4 points. 
> What do you think about that?
> Thanks, 
> Gerd.

To me the first point has the same ambiguity - do we perform constant-folding before or after checking if the expression depends on undefined?

However, I liked the use in the final example in your comment - I am just unsure if the semantics are specified.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 24 Jan 2014 13:49 UTC**
Thanks again. 
But in this proposal, parameters with undefined are only removed in annotations, not in usual equations. They simply stay there. So it doesn't matter whether you make constant folding before or after that.
Gerd.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 24 Jan 2014 13:55 UTC**
Replying to [comment:7 kurzbach]:
> Thanks again. 
> But in this proposal, parameters with undefined are only removed in annotations, not in usual equations. They simply stay there. So it doesn't matter whether you make constant folding before or after that.
> Gerd.

Oh, just removing them in annotations. 
Sorry, I didn't understand that restriction; I thought it was modifiers in general.

I agree that it seems like a good idea - will need to analyze it in more detail.

**Comment by modelica-trac-importer** [comment]
**Comment by volker&#x2e;beuter on 25 Jan 2014 14:29 UTC**
Replying to [comment:7 kurzbach]:
> But in this proposal, parameters with undefined are only removed in annotations, not in usual equations. They simply stay there. So it doesn't matter whether you make constant folding before or after that.

At least in the previous posts the discussion was not only about annotations, but at least also about definition equations of variables. In this case you _have_ the problem of constant folding, i.e. lazy evaluation.

But generally I see problems when undefining a variable is (recursively) passed to other variables: In most cases undefining a variable will not be done for removal but in order to determine it in some other way like an initial equation. Consider:

```mo
model standard
  parameter Real density = 7700;
  parameter Real mass = density*r^3;
  ...
equation
  F = mass*der(v);
  ...
end standard;
```

Now we build a variant by undefining  `density` and specifying an initial equation for it:
 
```mo
model variant
  extends standard(density(fixed = false) = undefined);
  ...
initial equation
  foo(density) = 0.0;
  ...
end variant;
```

If undefining `density` in `standard` also undefines `mass` (because it depends on `density` the model `variant` is missing an equation for `density` and we'd have to add the same definition equation in `variant` again.

Here I would rather propose that if some variable depending on an undefined variable is to be undefined also, the user has to do this manually and if a variable is still dependent on an undefined variable after adding new (initial) equations or algorithms this is an error.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 27 Jan 2014 06:05 UTC**
Sorry I did a mistake: The special handling of undefined described in point (1) of the 4 rules applies only to annotations.  Other modifications are seen as usual equations and left unchanged.
So `mass` in the example above will not be undefined in the `model variant` and the same equation will be used. See also the example after the points in the discussion.

Concerning constant folding in annotations:  Modifications in annotations are removed if they literally contain references to variables with undefined modification before any expression evaluation takes place, because `undefined` is not an expression and cannot be evaluated.
So maybe you are right. This handling has to be the first step.

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 12 Jun 2014 14:53 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 12 Jun 2014 15:15 UTC**
Here is another use case for undefined:

All the machines in Modelica.Electrical.Machines have an optional inertia Js for their stators. This inertia need only to be defined if the machine support is enabled (useSupport = true). If the support is enabled, the user should be forced to provide a value for Js, because it is not possible to give a meaning default. In order to avoid (wrong) error messages in Modelica tools, currently a dummy default value is given as start value:
```
parameter Real Jr;
parameter Boolean useSupport=false,
parameter Real Js(start=Jr) annotation(Dialog(enable=useSupport));
```
and Js is only "used" if useSupport=true (otherwise, there is no active equation that contains Js). This approach is not nice, because a tool may or may not give a warning/error if Js is not modified. To be on the safe side, a user needs to provide a value for Js, even if it is not needed and the user does not know it (so he/she just provides a dummy value). However, when changing to useSupport=true, the user may forget to set a meaningful value for Js and then the simulation is wrong (without noticing).

With an "undefined" keyword a much better solution is possible: 
```
parameter Real Jr;
parameter Boolean useSupport=false,
parameter Real Js = undefined annotation(Dialog(enable=useSupport));
```
If the support is not enabled, parameter Js is just removed from the model (as it should be, and Js can also not be set at run time).
If the support is enabled, then the user has to provide a value for Js, because otherwise the translator will give an error (Js = undefined, but Js is used in an equation).

So "undefined" allows a "clean" solution.

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 14 Jun 2014 14:41 UTC**
There is a new version of the MCPI document in the SVN.
It is significantly extended and reworked. I've added corrections and results from discussions in this ticket, clarifications and a discussion of the other proposals. 

Please use this a basis for further discussions.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 19 Dec 2014 07:01 UTC**
Ticket retargeted after milestone closed

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 18 Sep 2015 10:14 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 18 Sep 2015 10:22 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 18 Sep 2015 10:23 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 2 Dec 2015 08:31 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 25 Mar 2018 17:58 UTC**
This ticket is closed because there is a new one (#2241) discussing version 2 of MCP-0009.
Also because it containts to many side tracks, which are not in the sense of the original idea and complicate the understanding.
